### PR TITLE
Use int32 in Go interfaces that map to protobufs using int32

### PR DESCRIPTION
**Pull request by julienp**
We use plain `int`, which can be 32 or 64 bit based on the system, in some of the Go interfaces that map to protobuf using int32. Update the Go interfaces to use int32 to prevent potential integer overflow.

These potential integer conversion overflows are flagged by the latest golangci-lint. Once this PR is merged, we can upgrade our version of golangci-lint, which will unblock us from using Go 1.23 in CI.

This will require some small updates in providers, mostly removing casts `int32(...)`.

### CODE REVIEW

File: sdk/go/pulumi/context.go [hunk 1]
```diff
@@ -2378,8 +2379,16 @@ func (ctx *Context) getSourcePosition(skip int) *pulumirpc.SourcePosition {
 	for i := range elems {
 		elems[i] = url.PathEscape(elems[i])
 	}
+	var line int32
+	if frame.Line <= math.MaxInt32 {
+		//nolint:gosec
+		line = int32(frame.Line)
+	} else {
+		// line is out of range for int32, that's a long sourcefile!
+		line = 0
```
Comment (Frassle) [inline sdk/go/pulumi/context.go:2388]: Should we do -1 here to make it clear something is off?

**Review by Frassle — APPROVED** [review]

File: pkg/cmd/pulumi/up.go [hunk 2]
```diff
@@ -40,7 +40,8 @@ import (
 
 // The default number of parallel resource operations to run at once during an update, if --parallel is unset.
 // See https://github.com/pulumi/pulumi/issues/14989 for context around the cpu * 4 choice.
-var defaultParallel = runtime.NumCPU() * 4
+var defaultParallel = int32(runtime.NumCPU()) * 4 //nolint:gosec // NumCPU is an int32 internally,
+//                                                                  but the NumCPU function returns an int.
```
Comment (tgummerer) [inline pkg/cmd/pulumi/up.go:44]:
```suggestion
//nolint:gosec // NumCPU is an int32 internally, but the NumCPU function returns an int.
var defaultParallel = int32(runtime.NumCPU()) * 4
```

Very minor nitpick.  If you prefer the other way that's also fine with me :)

File: sdk/go/pulumi/run.go [hunk 3]
```diff
@@ -184,7 +184,7 @@ func getEnvInfo() RunInfo {
 		Stack:            os.Getenv(EnvStack),
 		Config:           config,
 		ConfigSecretKeys: configSecretKeys,
-		Parallel:         parallel,
+		Parallel:         int32(parallel), //nolint:gosec // guarded by strconv.ParseInt
```
Comment (tgummerer) [inline sdk/go/pulumi/run.go:187]: Should we cast it at the `ParseInt`?  Also what happens if that fails, we don't seem to have any error handling there :grimacing: It's probably not critical as I think if the syntax is wrong we'll get 0 here and it's going to be fine through the rest of the program.

**Review by tgummerer — APPROVED** [review]

File: sdk/go/pulumi/run.go (hunk 3 above)
Comment (Frassle) [reply sdk/go/pulumi/run.go:187]: Why I wondered if ParseInt32 would be helpful

**Pull request merged**